src/library/blast/arith/heuristic.cpp

```diff
@@ -0,0 +1,552 @@
+            return list<poly>(linearize(A, new_rhs, false, [=]() {
+                        return get_app_builder().mk_app(get_ordered_arith_zero_le_of_le_name(), h.get_self());
+                    }));
+        } else if (is_eq(h.get_type(), A, lhs, rhs) && type_ok(A)) {
+            expr new_rhs1 = get_app_builder().mk_add(A, rhs, get_app_builder().mk_neg(A, lhs));
+            expr new_rhs2 = get_app_builder().mk_add(A, lhs, get_app_builder().mk_neg(A, rhs));
+            poly p1 = linearize(A, new_rhs1, false, [=]() {
+                    return get_app_builder().mk_app(get_ordered_arith_zero_le_of_eq1_name(), h.get_self());
+                });
+            poly p2 = linearize(A, new_rhs2, false, [=]() {
+                    return get_app_builder().mk_app(get_ordered_arith_zero_le_of_eq2_name(), h.get_self());
+                });
+            return list<poly>({p1, p2});
+        } else {
+            return list<poly>();
+        }
+    }
+};
+
+/* Poly-pot */
+class poly_pot {
+    list<poly> m_positives;
+    list<poly> m_negatives;
+public:
+    void insert(poly const & p) {
+        if (p.is_positive()) {
+            m_positives = cons(p, m_positives);
+        } else {
+            m_negatives = cons(p, m_negatives);
+        }
+    }
+    list<poly> get_positives() { return m_positives; }
+    list<poly> get_negatives() { return m_negatives; }
+};
+
+struct heuristic_branch_extension : public branch_extension {
+private:
+    // TODO(dhs): will this cause problems during rebalancing?
+    // Should I wrap this class and make a cheap copy?
+    typedef rb_map<std::vector<atom>, poly_pot, atoms_quick_cmp> poly_pot_map;
+    linearizer        m_linearizer;
+    poly_pot_map      m_poly_pot_map;
+    list<poly>        m_todo;
+public:
+    heuristic_branch_extension() {}
+    heuristic_branch_extension(heuristic_branch_extension const & ext):
+        m_linearizer(ext.m_linearizer), m_poly_pot_map(ext.m_poly_pot_map), m_todo(ext.m_todo) {}
+    virtual branch_extension * clone() override { return new heuristic_branch_extension(*this); }
+
+    void put_todo(buffer<poly> const & todo) { m_todo = to_list(todo); }
+    void get_todo(buffer<poly> & todo) {
+        to_buffer(m_todo, todo);
+        m_todo = list<poly>();
+    }
+
+    list<poly> linearize(hypothesis_idx hidx) { return m_linearizer.linearize(hidx); }
+
+    poly_pot insert_poly_into_pot(poly const & p) {
+        poly_pot const * pot_p = m_poly_pot_map.find(p.get_major_idx());
+        poly_pot pot = (pot_p ? *pot_p : poly_pot());
+        pot.insert(p);
+        m_poly_pot_map.insert(p.get_major_idx(), pot);
+        return pot;
+    }
+};
+
+static heuristic_branch_extension & get_heuristic_extension() {
+    return static_cast<heuristic_branch_extension&>(curr_state().get_extension(g_ext_id));
+}
+
+/* HEURISTIC function */
+class heuristic_fn {
+    heuristic_branch_extension &        m_ext;
+    buffer<poly>                  m_todo;
```

Comment:
poly is not a small object. We will have to perform copies whenever inserting elements in the buffer.